rust/parser/mod.rs

```diff
@@ -20,5 +20,598 @@
+        NaiveDateTime::parse_from_str(date_time_text, "%Y-%m-%dT%H:%M").ok()
+    }
+}
+
+fn get_date_time(date_time: Rc<TerminalNode>) -> ParserResult<NaiveDateTime> {
+    parse_date_time(&date_time.get_text())
+        .ok_or(ILLEGAL_GRAMMAR.format(&[date_time.get_text().as_str()]))
+}
+
+fn get_var(var: Rc<TerminalNode>) -> UnboundVariable {
+    let name = &var.symbol.get_text()[1..];
+    if name == "_" {
+        UnboundVariable::anonymous()
+    } else {
+        UnboundVariable::named(String::from(name))
+    }
+}
+
+fn get_isa_constraint(
+    _isa: Rc<TerminalNode>,
+    ctx: Rc<Type_ContextAll>,
+) -> ParserResult<IsaConstraint> {
+    match visit_type(ctx)? {
+        Type::Unscoped(label) => Ok(IsaConstraint::from(label)),
+        Type::Variable(var) => Ok(IsaConstraint::from(var)),
+        _ => Err(ILLEGAL_STATE.format(&[line!().to_string().as_str()])),
+    }
+}
+
+fn get_role_player_constraint(ctx: Rc<Role_playerContext>) -> ParserResult<RolePlayerConstraint> {
+    let player = get_var(ctx.player().unwrap().VAR_().unwrap());
+    if let Some(type_) = ctx.type_() {
+        match visit_type(type_)? {
+            Type::Unscoped(label) => Ok(RolePlayerConstraint::from((label, player))),
+            Type::Variable(var) => Ok(RolePlayerConstraint::from((var, player))),
+            _ => Err(ILLEGAL_STATE.format(&[line!().to_string().as_str()])),
+        }
+    } else {
+        Ok(RolePlayerConstraint::from(player))
+    }
+}
+
+fn get_role_players(ctx: Rc<RelationContext>) -> ParserResult<Vec<RolePlayerConstraint>> {
+    (0..)
+        .map_while(|i| ctx.role_player(i))
+        .map(get_role_player_constraint)
+        .collect()
+}
+
+pub fn visit_eof_query(ctx: Rc<Eof_queryContext>) -> ParserResult<Query> {
+    visit_query(ctx.query().unwrap())
+}
+
+fn visit_eof_queries(ctx: Rc<Eof_queriesContext>) -> ParserResult<Vec<Query>> {
+    (0..).map_while(|i| ctx.query(i)).map(visit_query).collect()
+}
+
+fn visit_eof_pattern(ctx: Rc<Eof_patternContext>) -> ParserResult<Pattern> {
+    visit_pattern(ctx.pattern().unwrap())
+}
+
+fn visit_eof_patterns(ctx: Rc<Eof_patternsContext>) -> ParserResult<Vec<Pattern>> {
+    visit_patterns(ctx.patterns().unwrap())
+}
+
+fn visit_eof_definables(ctx: Rc<Eof_definablesContext>) -> ParserResult<Vec<Definable>> {
+    let definables_ctx = ctx.definables().unwrap();
+    (0..)
+        .map_while(|i| definables_ctx.definable(i))
+        .map(visit_definable)
+        .collect()
+}
+
+fn visit_eof_variable(ctx: Rc<Eof_variableContext>) -> ParserResult<Variable> {
+    visit_pattern_variable(ctx.pattern_variable().unwrap())
+}
+
+fn visit_eof_label(ctx: Rc<Eof_labelContext>) -> ParserResult<String> {
+    Ok(ctx.label().unwrap().get_text())
+}
+
+fn visit_eof_schema_rule(ctx: Rc<Eof_schema_ruleContext>) -> ParserResult<()> {
+    visit_schema_rule(ctx.schema_rule().unwrap())
+}
+
+fn visit_query(ctx: Rc<QueryContext>) -> ParserResult<Query> {
+    if let Some(query_match) = ctx.query_match() {
```

Comment:
Now that `ParserResult` is a `Result`, are we able to collapse this into something like `ctx.query_match().and_then(|x| visit_query_match(x)).and_then(|x| x.into_query()).map_err(|| ILLEGAL_GRAMMAR.format(...))`? (pseudo-Rust, do not trust my ability to type valid Rust in a text editor with no compiler)

Reply:
Not really, this will end up being
```rust
    if let Some(query_match) = ctx.query_match() {
        Ok(visit_query_match(query_match)?.into_query())
    } else if let Some(query_insert) = ctx.query_insert() {
        Ok(visit_query_insert(query_insert)?.into_query())
    } else ...
```
so not really chainable afaics.